### ISSUE TITLE
CSRF Improvement proposal

### ISSUE THREAD

**Issue by SamMousa** [yiisoft/yii2#16214]
I propose storing the CSRF token in LocalStorage instead of a `meta` tag.
Inside each page we can then add the following javascript snippet to auto update hidden form input elements in case the CSRF token changes from another window:
```js
// Created instead of the meta tag:
localStorage.setItem('yiiCsrfToken', "xyz");

// Allows this window to be notified if the csrf token changes, note that this is pretty cheap performance wise.
window.addEventHandler('storage', function(e) {
    if (e.key === 'yiiCsrfToken') {
        // do some stuff here 
        $('input[name=_csrf]').val(e.newValue);
    } 
});
```

This will allow for a better session resumption user experience. 

This does not tackle the session renewal itself, it just makes sure that when a CSRF token is renewed other windows are kept in sync.

Without any evidence to back it up, it could also resolve some issues like #16200 and #9504.

**Comment by bizley** [comment]
I would be cautious about this. Browsers sometimes behave strange with local storage. See [this article](https://michalzalecki.com/why-using-localStorage-directly-is-a-bad-idea/) for example. And even if everything is fixed with current browsers' versions we should think about the previous ones. And this would lead to `meta` solution as a fallback anyway.

**Comment by SamMousa** [comment]
Fair enough, but that's more of a **when** than an **if**.
https://caniuse.com/#search=localstorage

The fallback would be to set it somewhere in javascript. (Since the meta token is only used by javascript anyway and we have a Yii object in the window).

Note that LocalStorage is a few years old already, it's not brand new stuff.

**Comment by bizley** [comment]
True but as you can see even canIUse reports known issues with it, mainly with IE, iOS and Safari.

**Comment by SamMousa** [comment]
Well regardless, I'm okay with having a fallback. 
This fallback could then store it directly in `yii.js` instead of using the meta tag as storage.
Since any consumer will use it either via javascript or via hidden form inputs (which of course will also be updated).

**Comment by samdark** [comment]
OK. That makes sense.

**Comment by SamMousa** [comment]
I have thought some more on this, I'd like to take this opportunity to remove CSRF from `Request` and `Controller` and implement it as a separate component and filter.

1. `CsrfFilter extends ActionFilter`.
2. `Csrf extends Component`.
3. `Html` will use the `csrf` component to render the CSRF tag inside forms.
4. `Csrf` component will optionally register the CSRF renewal script.

### Pros:
- Csrf becomes just another filter instead of a specific case.
- Request component is no longer responsible for CSRF.
- Controller component is no longer doing CSRF.

### Cons
- People should take care to use `parent::behaviors()` in their controllers to make sure CSRF is included. 

### Feedback needed
One thing I'm not sure about is how / if a component should be registering client side assets.
- It could get the application view, but there's no guarantee that that's the view that's being used.
- It could use class level events, but that doesn't feel right.


Discuss please!

**Comment by samdark** [comment]
I like the idea.

> One thing I'm not sure about is how / if a component should be registering client side assets.

No, it should not.

**Comment by SamMousa** [comment]
So then we offer a bundle that people can include in their page or AppAsset?

**Comment by SamMousa** [comment]
That bundle could even add the meta tag as well, removing the need for the HTML csrf meta tag function

**Comment by samdark** [comment]
I don't think we need any JS there at all except in the docs pointing the fact that there's meta-tag with CSRF token that could be read easily via JS.

**Comment by SamMousa** [comment]
I'd like to offer the JS for syncing between browser windows as an asset as well; it doesn't need to be included by default of course.

**Comment by samdark** [comment]
Localstorage. Right. Makes sense,

**Comment by SamMousa** [comment]
@bizley I don't think so. The only "extra" thing we add is CSRF propagation via localStorage.
Since localStorage is domain specific so there's no risk of XSS.

**Comment by bizley** [comment]
Isn't it the case when included 3rd party JS can read and change localStorage? I'm not 100% sure, just asking.

**Comment by SamMousa** [comment]
Well yes.. included 3rd party JS can read localStorage, but they can also read your DOM, send requests to the applications' domain with cookies...
Basically if you include unsafe JS it's not XSS, it's stupidity. Unfortunately there are few known mitigation againsts that :dancing_men:
